### PR TITLE
SOS bugfixes: PIC-1157, PIC-1158

### PR DESCRIPTION
**Pull request by michaelk-stratoscale**
@rom-stratoscale @eliran-stratoscale @noam-stratoscale

### CODE REVIEW

File: inaugurator/main.py [hunk 1]
```diff
@@ -44,6 +45,8 @@
 parser.add_argument("--inauguratorDontReadSmartData", action="store_true", default=False)
 parser.add_argument("--inauguratorDontFailOnFailedDisk", action="store_true", default=False)
 parser.add_argument("--inauguratorCleanupUpperPercentageThreshold", type=int, default=65)
+parser.add_argument("--inauguratorWipeOldInauguratorInstallations", type=bool, action="store_true",
```
Comment (rom-stratoscale) [inline inaugurator/main.py]: lets have the default to false please!!!

File: inaugurator/ceremony.py [hunk 2]
```diff
@@ -313,20 +314,43 @@ def _additionalDownload(self, destination):
...
+        if self._args.inauguratorTargetDeviceCandidate is not None:
+            logging.info("Explicit target devices indicated: %(deviceCandidates)s",
+                         dict(deviceCandidates=self._args.inauguratorTargetDeviceCandidate))
+            candidates = self._args.inauguratorTargetDeviceCandidate
+        elif self._args.inauguratorTargetDeviceType is not None:
             logging.info("Searching for target devices of type %(deviceType)s",
                          dict(deviceType=self._args.inauguratorTargetDeviceType))
             device = self._storageDevices.findFirstDeviceOfType(self._args.inauguratorTargetDeviceType,
                                                                 self._talkToServer)
+            if not device:
```
Comment (rom-stratoscale) [inline inaugurator/ceremony.py]: first add logging

File: inaugurator/ceremony.py [hunk 3]
```diff
@@ -313,20 +314,43 @@ def _additionalDownload(self, destination):
...
+
+    def _setTargetDevice(self):
+        if self._args.inauguratorTargetDeviceCandidate is not None:
+            logging.info("Explicit target devices indicated: %(deviceCandidates)s",
+                         dict(deviceCandidates=self._args.inauguratorTargetDeviceCandidate))
+            candidates = self._args.inauguratorTargetDeviceCandidate
+        elif self._args.inauguratorTargetDeviceType is not None:
             logging.info("Searching for target devices of type %(deviceType)s",
                          dict(deviceType=self._args.inauguratorTargetDeviceType))
             device = self._storageDevices.findFirstDeviceOfType(self._args.inauguratorTargetDeviceType,
```
Comment (rom-stratoscale) [inline inaugurator/ceremony.py]: why first? and not every device - and then assert? I think it would be safer, no?

Comment (rom-stratoscale) [inline inaugurator/ceremony.py]: please notice that findFirstDeviceOfType raises and exception.

File: inaugurator/ceremony.py [hunk 4]
```diff
@@ -313,20 +314,43 @@ def _additionalDownload(self, destination):
...
+                raise Exception("No devices found with label %s" % self._args.inauguratorTargetDeviceLabel)
+
         self._targetDevice = targetdevice.TargetDevice.device(candidates)
-        self._createPartitionTable()
-        logging.info("Partitions created")
-        self._mountOp = mount.Mount()
-        assert self._bootPartitionPath is not None, "Please initialize boot partition path first"
-        self._mountOp.setBootPartitionPath(self._bootPartitionPath)
+
+    def _getDevicesWithLabel(self, label):
```
Comment (rom-stratoscale) [inline inaugurator/ceremony.py]: I guess it should be part of storagedevices.py file, no?

File: inaugurator/ceremony.py [hunk 5]
```diff
@@ -313,20 +314,43 @@ def _additionalDownload(self, destination):
...
+        candidateList = []
+        output = sh.run("blkid")
+        for line in output.splitlines():
+            entry = line.strip().split(":", 1)
```
Comment (rom-stratoscale) [inline inaugurator/ceremony.py]: checking on a live cluster:
/dev/mapper/inaugurator-temp: LABEL="temp" UUID="de711c76-99a2-4825-b9c6-956f8a9e58bb" TYPE="ext4" 
/dev/mapper/inaugurator-runtime_swap: LABEL="runtime_swap" UUID="6862f17f-ce77-41d6-b438-dbbcff932da7" TYPE="ext4" 
/dev/mapper/inaugurator-data: LABEL="data" UUID="11e5df1e-c6c9-4e11-82c9-820e61b03e6c" TYPE="ext4" 
/dev/mapper/docker-253:2-1048580-25257c2369d024e32045b64a2dc121933604d26c4daafa5a10112d12976a4a74: UUID="27fe0dc5-a035-4617-b247-dc94f6b926d7" TYPE="xfs" 
/dev/mapper/docker-253:2-1048580-26a46febce96d1ecf2c4e6637d1f4747beced6bf43e17fd4c754799bf82eb783: UUID="27fe0dc5-a035-4617-b247-dc94f6b926d7" TYPE="xfs" 


will ur split work ? Or you should not see docker mounts there?

File: inaugurator/ceremony.py [hunk 6]
```diff
@@ -313,20 +314,43 @@ def _additionalDownload(self, destination):
...
+            if len(entry) == 2 and " LABEL=\"%s\"" % label in entry[1]:
+                candidateList.append(entry[0].strip())
```
Comment (rom-stratoscale) [inline inaugurator/ceremony.py]: please log the matched line

File: inaugurator/ceremony.py [hunk 7]
```diff
@@ -313,20 +314,43 @@ def _additionalDownload(self, destination):
...
                                                                 self._talkToServer)
+            if not device:
+                raise Exception("No %s devices found" % self._args.inauguratorTargetDeviceType)
             candidates = [device]
-        else:
-            candidates = self._args.inauguratorTargetDeviceCandidate
+        elif self._args.inauguratorTargetDeviceLabel is not None:
+            logging.info("Searching for target devices with label %(deviceLabel)s",
+                         dict(deviceLabel=self._args.inauguratorTargetDeviceLabel))
+            candidates = self._getDevicesWithLabel(self._args.inauguratorTargetDeviceLabel)
```
Comment (rom-stratoscale) [inline inaugurator/ceremony.py]: if u get 2, what will happen?

**Comment by rom-stratoscale** [comment]
I still think that you can cause a creation of another BOOT partition - once u get into partitiontable.py, and there u'll find a '_findMismatch' - you'll go to _create logic that willl create BOOT. @eliran-stratoscale is this flow possible with current logic?

File: inaugurator/main.py (hunk 1 above)
Comment (michaelk-stratoscale) [inline inaugurator/main.py]: We must be backward compatible if we don't want to restart rackattack.

Comment (rom-stratoscale) [inline inaugurator/main.py]: anyhow we'll have to restart rackattack. And I rather restart rackattack than by accident erase another customer

**Comment by noam-stratoscale** [comment]
i dont know if for this task but we should really add loggings as the code of inaugurator is very hard to debug with eyes

File: inaugurator/ceremony.py (hunk 3 above)
Comment (eliran-stratoscale) [inline inaugurator/ceremony.py]: What do you mean by 'every device'? @rom-stratoscale

File: inaugurator/ceremony.py (hunk 4 above)
Comment (eliran-stratoscale) [inline inaugurator/ceremony.py]: Please use the already existing code in partitiontable.py and don't rewrite this.

File: inaugurator/main.py [hunk 8]
```diff
@@ -32,6 +32,7 @@
 parser.add_argument("--inauguratorDownload", nargs='+', default=[])
 parser.add_argument("--inauguratorIgnoreDirs", nargs='+', default=[])
 parser.add_argument("--inauguratorTargetDeviceCandidate", nargs='+')
+parser.add_argument("--inauguratorTargetDeviceLabel", default="BOOT")
```
Comment (eliran-stratoscale) [inline inaugurator/main.py]: Add a 'help' argument to the three device detection methods that says that those 3 are mutually exclusive.

File: inaugurator/partitiontable.py [hunk 9]
```diff
@@ -327,15 +329,19 @@ def verify(self):
...
+                self._wipeOtherPartitionsWithSameVolumeGroup()
+                self._wipeOtherPartitionsWithBootLabel()
             return
         self._create()
         for retry in xrange(5):
             mismatch = self._findMismatch()
             if mismatch is None:
-                self._wipeOtherPartitionsWithSameVolumeGroup()
-                self._wipeOtherPartitionsWithBootLabel()
+                if self._wipeOldInstallations:
```
Comment (eliran-stratoscale) [inline inaugurator/partitiontable.py]: I think it's enough code to be eligible to its own method.

File: inaugurator/main.py [hunk 10]
```diff
@@ -44,6 +45,8 @@
 parser.add_argument("--inauguratorDontReadSmartData", action="store_true", default=False)
 parser.add_argument("--inauguratorDontFailOnFailedDisk", action="store_true", default=False)
 parser.add_argument("--inauguratorCleanupUpperPercentageThreshold", type=int, default=65)
+parser.add_argument("--inauguratorWipeOldInauguratorInstallations", type=bool, action="store_true",
+                    optional=False)
```
Comment (eliran-stratoscale) [inline inaugurator/main.py]: Consider whether this means a new major version.

File: inaugurator/ceremony.py (hunk 7 above)
Comment (eliran-stratoscale) [inline inaugurator/ceremony.py]: After talking with Rom we've realized that this validation (no more than 1 BOOT partition) should not only happen under this option (inauguratorTargetDeviceLabel), but at any option. So, consider either:
* doing this validation with 1 tab less (validate that there are no BOOT labels on devices which are not the target device).
* Adding partitiontable.py - when verifying the partition table after creation, then if it's set to NOT wipe other boot partitions, then also raise an exception there.
* Both of the above.

File: inaugurator/main.py (hunk 8 above)
Comment (eliran-stratoscale) [inline inaugurator/main.py]: This should not have a default value. I think that the user should explicitly use one of the 3 flags needed (and if not, ceremony.py will throw an exception).

File: inaugurator/main.py [hunk 11]
```diff
@@ -32,6 +32,7 @@
 parser.add_argument("--inauguratorDownload", nargs='+', default=[])
 parser.add_argument("--inauguratorIgnoreDirs", nargs='+', default=[])
 parser.add_argument("--inauguratorTargetDeviceCandidate", nargs='+')
+parser.add_argument("--inauguratorTargetDeviceLabel", default="BOOT")
 parser.add_argument("--inauguratorTargetDeviceType", default="SSD")
```
Comment (eliran-stratoscale) [inline inaugurator/main.py]: State in the 'help' section that it's permissive (first device of type).

File: inaugurator/ceremony.py (hunk 2 above)
Comment (michaelk-stratoscale) [inline inaugurator/ceremony.py]: Fixed

File: inaugurator/ceremony.py (hunk 4 above)
Comment (michaelk-stratoscale) [inline inaugurator/ceremony.py]: Used partitiontable

File: inaugurator/main.py (hunk 8 above)
Comment (michaelk-stratoscale) [inline inaugurator/main.py]: Done

File: inaugurator/partitiontable.py (hunk 9 above)
Comment (michaelk-stratoscale) [inline inaugurator/partitiontable.py]: Done

File: inaugurator/ceremony.py (hunk 5 above)
Comment (michaelk-stratoscale) [inline inaugurator/ceremony.py]: No worries

**Comment by michaelk-stratoscale** [comment]
@rom-stratoscale @noam-stratoscale @eliran-stratoscale 
Fixed. Please have another go.

File: inaugurator/ceremony.py [hunk 12]
```diff
@@ -313,20 +314,32 @@ def _additionalDownload(self, destination):
...
                                                                 self._talkToServer)
             candidates = [device]
-        else:
-            candidates = self._args.inauguratorTargetDeviceCandidate
+        elif self._args.inauguratorTargetDeviceLabel is not None:
+            logging.info("Searching for target devices with label %(deviceLabel)s",
+                         dict(deviceLabel=self._args.inauguratorTargetDeviceLabel))
+            candidates = partitiontable.PartitionTable.getDevicesWithLabel(self._args.inauguratorTargetDeviceLabel)
+            if not candidates:
+                raise Exception("No devices found with label %s" % self._args.inauguratorTargetDeviceLabel)
```
Comment (michaelk-stratoscale) [inline inaugurator/ceremony.py]: Double checking, but ok...

**Comment by rom-stratoscale** [comment]
@eliran-stratoscale @michaelk-stratoscale there are both logging and print.... please unite. Also, @eliran-stratoscale we have a ticket to write logs to file (if possible)

**Pull request merged**